servermaster/job_fsm.go

```diff
@@ -0,0 +1,153 @@
+package servermaster
+
+import (
+	"sync"
+
+	"github.com/hanfei1991/microcosm/lib"
+	"github.com/hanfei1991/microcosm/pkg/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/zap"
+)
+
+type jobHolder struct {
+	lib.WorkerHandle
+	*lib.JobMasterV2
+}
+
+// JobFsm manages state of all job masters, job master state forms a finite-state
+// machine. Note job master managed in JobFsm is in running status, which means
+// the job is not terminated or finished.
+//
+// ,-------.                   ,-------.            ,-------.
+// |WaitAck|                   |Online |            |Pending|
+// `---+---'                   `---+---'            `---+---'
+//     |                           |                    |
+//     | Master                    |                    |
+//     |  .OnWorkerOnline          |                    |
+//     |-------------------------->|                    |
+//     |                           |                    |
+//     |                           |                    |
+//     |                           | Master             |
+//     |                           |   .OnWorkerOffline |
+//     |                           |------------------->|
+//     |                           |                    |
+//     |                           |                    |
+//     |                           | Master             |
+//     |                           |   .CreateWorker    |
+//     |<-----------------------------------------------|
+//     |                           |                    |
+//     |                           |                    |
+//     | Master                    |                    |
+//     |  .OnWorkerDispatched      |                    |
+//     |  (with error)             |                    |
+//     |----------------------------------------------->|
+//     |                           |                    |
+//     |                           |                    |
+//     |                           |                    |
+type JobFsm struct {
+	JobStats
+
+	jobsMu      sync.RWMutex
+	pendingJobs map[string]*lib.JobMasterV2
```

Comment:
Why not use lib.workerID as key type?

Reply:
Changed to `MasterID`, actually they alias to string actually.